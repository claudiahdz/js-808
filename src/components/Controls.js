import React from 'react';

const Controls = ({ play, stop, generateBeat, changeTempo, clear, sequences }) => {
  return (
    <div className="buttons">
      <div className="button" onClick={play}>Play</div>
      <div className="button" onClick={stop}>Stop</div>
      <select className="button" onChange={generateBeat}>
        <option value={0}>4 on the floor</option>
        { sequences.slice(1).map((elem, i) => <option value={++i} key={i}>Sequence {i}</option>) }
      </select>
      <select className="button" onChange={changeTempo}>
        <option value={60}>60 BMP</option>   
        <option value={120}>120 BMP</option>      
        <option value={180}>180 BMP</option>      
      </select> 
      <div className="button" onClick={clear}>Clear</div>
    </div>    
  )
}

export default Controls