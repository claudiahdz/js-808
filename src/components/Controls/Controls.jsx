import React from 'react'
import PropTypes from 'prop-types'
import { sequences } from 'config/config'

const Controls = ({ status, play, stop, pause, clear, changeBeat, changeTempo, tempo } ) => {
  const isPlaying = (status === 'play')

  return (
    <div className="buttons">
      <div className="button" onClick={isPlaying ? pause : play}>{isPlaying ? 'Pause' : 'Play'}</div>
      <div className="button" onClick={stop}>Stop</div>
      <select className="button" onChange={(e) => changeBeat(sequences[e.target.value].map(r => r.slice()))}>
        <option value={0}>4 on the floor</option>
        { sequences.slice(1).map((elem, i) => <option value={++i} key={i}>Sequence {i}</option>) }
      </select>
      <input type="range" min="60" max="180" value={tempo} onChange={(e) => changeTempo(e.target.value)} /> 
      <span>{tempo} BMP</span>
      <div className="button" onClick={clear}>Clear</div>
    </div>    
  )
}

Controls.propTypes = {
  status: PropTypes.string.isRequired,
  tempo: PropTypes.number.isRequired,
  play: PropTypes.func.isRequired,
  stop: PropTypes.func.isRequired,
  pause: PropTypes.func.isRequired,
  clear: PropTypes.func.isRequired,
  changeBeat: PropTypes.func.isRequired,
  changeTempo: PropTypes.func.isRequired
}

export default Controls
