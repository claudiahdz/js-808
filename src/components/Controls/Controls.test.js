import React from 'react'
import { shallow, render, mount } from 'enzyme'
import Controls from './Controls'

describe('Controls', () => {
  const defaultProps = {
    status: 'stop',
    play: () => {},
    stop: () => {},
    pause: () => {},
    clear: () => {},
    changeBeat: () => {},
    changeTempo: () => {},
    tempo: 60,    
  }

  it('renders the component correctly', () => {
    const controls = shallow(<Controls {...defaultProps} />)
    expect(controls).toMatchSnapshot()
  })

  it('plays beat', () => {
    const props = { 
      ...defaultProps,
      play: jest.fn()
    }
    const controls = shallow(<Controls {...props} />)
    const button = controls.find('.button').first()

    button.simulate('click')
    expect(props.play).toHaveBeenCalled()
  })

  it('pauses beat', () => {
    const props = { 
      ...defaultProps,
      status: 'play',
      pause: jest.fn()
    }
    const controls = shallow(<Controls {...props} />)
    const button = controls.find('.button').first()

    button.simulate('click')
    expect(props.pause).toHaveBeenCalled()
  })

  it('stops beat', () => {
    const props = { 
      ...defaultProps,
      stop: jest.fn()
    }
    const controls = shallow(<Controls {...props} />)
    const button = controls.find('.button').at(1)

    button.simulate('click')
    expect(props.stop).toHaveBeenCalled()
  })

  it('changes tempo', () => {
    const props = { 
      ...defaultProps,
      changeTempo: jest.fn()
    }
    const e = { target: {value: 60} }

    const controls = shallow(<Controls {...props} />)
    const input = controls.find('input')
    input.first().simulate('change', e)

    expect(props.changeTempo).toHaveBeenCalled()
  })

  it('changes beat', () => {
    const props = { 
      ...defaultProps,
      changeBeat: jest.fn()
    }
    const e = { target: {value: 0} }

    const controls = shallow(<Controls {...props} />)
    const select = controls.find('select')
    select.first().simulate('change', e)

    expect(props.changeBeat).toHaveBeenCalled()
  })

  it('clears beat', () => {
    const props = { 
      ...defaultProps,
      clear: jest.fn()
    }
    const controls = shallow(<Controls {...props} />)
    const button = controls.find('.button').at(3)

    button.simulate('click')
    expect(props.clear).toHaveBeenCalled()
  })

})
