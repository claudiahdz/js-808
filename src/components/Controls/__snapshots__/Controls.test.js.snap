// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls renders the component correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className="buttons"
>
    <div
        className="button"
        onClick={[Function]}
    >
        Play
    </div>
    <div
        className="button"
        onClick={[Function]}
    >
        Stop
    </div>
    <select
        className="button"
        onChange={[Function]}
    >
        <option
            value={0}
        >
            4 on the floor
        </option>
        <option
            value={1}
        >
            Sequence 
            1
        </option>
        <option
            value={2}
        >
            Sequence 
            2
        </option>
        <option
            value={3}
        >
            Sequence 
            3
        </option>
    </select>
    <input
        max="180"
        min="60"
        onChange={[Function]}
        type="range"
        value={60}
    />
    <span>
        60
         BMP
    </span>
    <div
        className="button"
        onClick={[Function]}
    >
        Clear
    </div>
</div>,
  "nodes": Array [
    <div
      className="buttons"
>
      <div
            className="button"
            onClick={[Function]}
      >
            Play
      </div>
      <div
            className="button"
            onClick={[Function]}
      >
            Stop
      </div>
      <select
            className="button"
            onChange={[Function]}
      >
            <option
                  value={0}
            >
                  4 on the floor
            </option>
            <option
                  value={1}
            >
                  Sequence 
                  1
            </option>
            <option
                  value={2}
            >
                  Sequence 
                  2
            </option>
            <option
                  value={3}
            >
                  Sequence 
                  3
            </option>
      </select>
      <input
            max="180"
            min="60"
            onChange={[Function]}
            type="range"
            value={60}
      />
      <span>
            60
             BMP
      </span>
      <div
            className="button"
            onClick={[Function]}
      >
            Clear
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Controls
        changeBeat={[Function]}
        changeTempo={[Function]}
        clear={[Function]}
        pause={[Function]}
        play={[Function]}
        status="stop"
        stop={[Function]}
        tempo={60}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "changeBeat": [Function],
          "changeTempo": [Function],
          "clear": [Function],
          "pause": [Function],
          "play": [Function],
          "status": "stop",
          "stop": [Function],
          "tempo": 60,
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className="buttons"
>
          <div
                    className="button"
                    onClick={[Function]}
          >
                    Play
          </div>
          <div
                    className="button"
                    onClick={[Function]}
          >
                    Stop
          </div>
          <select
                    className="button"
                    onChange={[Function]}
          >
                    <option
                              value={0}
                    >
                              4 on the floor
                    </option>
                    <option
                              value={1}
                    >
                              Sequence 
                              1
                    </option>
                    <option
                              value={2}
                    >
                              Sequence 
                              2
                    </option>
                    <option
                              value={3}
                    >
                              Sequence 
                              3
                    </option>
          </select>
          <input
                    max="180"
                    min="60"
                    onChange={[Function]}
                    type="range"
                    value={60}
          />
          <span>
                    60
                     BMP
          </span>
          <div
                    className="button"
                    onClick={[Function]}
          >
                    Clear
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="buttons"
>
          <div
                    className="button"
                    onClick={[Function]}
          >
                    Play
          </div>
          <div
                    className="button"
                    onClick={[Function]}
          >
                    Stop
          </div>
          <select
                    className="button"
                    onChange={[Function]}
          >
                    <option
                              value={0}
                    >
                              4 on the floor
                    </option>
                    <option
                              value={1}
                    >
                              Sequence 
                              1
                    </option>
                    <option
                              value={2}
                    >
                              Sequence 
                              2
                    </option>
                    <option
                              value={3}
                    >
                              Sequence 
                              3
                    </option>
          </select>
          <input
                    max="180"
                    min="60"
                    onChange={[Function]}
                    type="range"
                    value={60}
          />
          <span>
                    60
                     BMP
          </span>
          <div
                    className="button"
                    onClick={[Function]}
          >
                    Clear
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Controls
    changeBeat={[Function]}
    changeTempo={[Function]}
    clear={[Function]}
    pause={[Function]}
    play={[Function]}
    status="stop"
    stop={[Function]}
    tempo={60}
/>,
}
`;
